train_and_eval/train/train_imagenet.py

```diff
@@ -0,0 +1,607 @@
+        if use_blurpool: apply_blurpool(model)
+
+        model = model.to(memory_format=ch.channels_last)
+        model = model.to(self.gpu)
+
+        if distributed:
+            model = ch.nn.parallel.DistributedDataParallel(model, device_ids=[self.gpu])
+
+        return model, scaler
+
+    @param('logging.log_level')
+    def train_loop(self, epoch, log_level):
+        model = self.model
+        model.train()
+        losses = []
+
+        lr_start, lr_end = self.get_lr(epoch), self.get_lr(epoch + 1)
+        iters = len(self.train_loader)
+        lrs = np.interp(np.arange(iters), [0, iters], [lr_start, lr_end])
+
+        iterator = tqdm(self.train_loader)
+        for ix, (images, target) in enumerate(iterator):
+            ### Training start
+            for param_group in self.optimizer.param_groups:
+                param_group['lr'] = lrs[ix]
+
+            self.optimizer.zero_grad(set_to_none=True)
+            with autocast():
+                output = self.model(images)
+                loss_train = self.loss(output, target)
+
+            self.scaler.scale(loss_train).backward()
+            self.scaler.step(self.optimizer)
+            self.scaler.update()
+            ### Training end
+
+            ### Logging start
+            if log_level > 0:
+                losses.append(loss_train.detach())
+
+                group_lrs = []
+                for _, group in enumerate(self.optimizer.param_groups):
+                    group_lrs.append(f'{group["lr"]:.3f}')
+
+                names = ['ep', 'iter', 'shape', 'lrs']
+                values = [epoch, ix, tuple(images.shape), group_lrs]
+                if log_level > 1:
+                    names += ['loss']
+                    values += [f'{loss_train.item():.3f}']
+
+                msg = ', '.join(f'{n}={v}' for n, v in zip(names, values))
+                iterator.set_description(msg)
+            ### Logging end
+
+        if log_level > 0:
+            loss = ch.stack(losses).mean().cpu()
+            assert not ch.isnan(loss), 'Loss is NaN!'
+            return loss.item()
+
+    @param('validation.lr_tta')
+    def val_loop(self, lr_tta):
+        model = self.model
+        model.eval()
+
+        with ch.no_grad():
+            with autocast():
+                for images, target in tqdm(self.val_loader):
+                    output = self.model(images)
+                    if lr_tta:
+                        output += self.model(ch.flip(images, dims=[3]))
+
+                    for k in ['top_1', 'top_5']:
+                        self.val_meters[k](output, target)
+
+                    loss_val = self.loss(output, target)
+                    self.val_meters['loss'](loss_val)
+
+        stats = {k: m.compute().item() for k, m in self.val_meters.items()}
+        [meter.reset() for meter in self.val_meters.values()]
+        return stats
+
+
+    @param('validation.lr_tta')
+    def val_loop_nesting(self, lr_tta):
```

Comment:
can we combine this method with `val_loop()` and have an internal flag to reduce code repetition?

Reply:
Ah it would get messy... there isn't actually much of code repeatition.

Reply:
no worries. relatively minor